rasterio/_warp.pyx

```diff
@@ -671,11 +696,38 @@ cdef class WarpedVRTReaderBase(DatasetReaderBase):
             self.dst_nodata, GDALGetRasterCount(hds), <const char **>c_warp_extras)
 
         try:
-            with nogil:
-                hds_warped = GDALAutoCreateWarpedVRT(
-                    hds, NULL, dst_crs_wkt, c_resampling,
-                    c_tolerance, psWOptions)
-            self._hds = exc_wrap_pointer(hds_warped)
+            if dst_width and dst_height and dst_transform:
+                # set up transform args (otherwise handled in GDALAutoCreateWarpedVRT)
+                try:
+                    hTransformArg = exc_wrap_pointer(
+                        GDALCreateGenImgProjTransformer(
+                            hds, NULL, NULL, dst_crs_wkt, True, 1.0, 0))
+                    if c_tolerance > 0.0:
+                        hTransformArg = exc_wrap_pointer(
+                            GDALCreateApproxTransformer(
+                                GDALGenImgProjTransform, hTransformArg, c_tolerance))
+                        pfnTransformer = GDALApproxTransform
+                        GDALApproxTransformerOwnsSubtransformer(hTransformArg, 1)
+
+                    log.debug("Created transformer and options.")
+
+                except:
+                    GDALDestroyApproxTransformer(hTransformArg)
+                    raise
+
+                psWOptions.hSrcDS = hds
+                psWOptions.pfnTransformer = pfnTransformer
+
```

Comment:
@mojodna needs a `psWOptions.pTransformerArg = hTransformArg` after line 719. I tried it and your script works :)